
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Advanced Dev Notes &#8212; plco-analysis v1.0.0 documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Methods Summary" href="Methods%20Summary.html" />
    <link rel="prev" title="User Vignette" href="User%20Vignette.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="advanced-dev-notes">
<h1>Advanced Dev Notes<a class="headerlink" href="#advanced-dev-notes" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview-of-make">
<h2>Overview of Make<a class="headerlink" href="#overview-of-make" title="Permalink to this headline">¶</a></h2>
<p>Existing pipelines within this project are in the language <code class="docutils literal notranslate"><span class="pre">make</span></code>; my condolences!
<code class="docutils literal notranslate"><span class="pre">make</span></code> lacks many of the features of modern workflow languages, so some of the usage
may be unusual or quirky, or lacking entirely in features you’d expect from <code class="docutils literal notranslate"><span class="pre">snakemake</span></code>
or other similar languages.</p>
<div class="section" id="rule-structures">
<h3>Rule Structures<a class="headerlink" href="#rule-structures" title="Permalink to this headline">¶</a></h3>
<p>If you’re not familiar with <code class="docutils literal notranslate"><span class="pre">make</span></code> at all, you should definitely skim the <a class="reference external" href="https://www.gnu.org/software/make/manual/html_node/index.html">make manual</a>
first. I’ll just cover some extremely brief thoughts on <code class="docutils literal notranslate"><span class="pre">make</span></code> usage in these pipelines.</p>
<ul class="simple">
<li><p>rules rely heavily on text replacement and substitution for building input files from outputs;
see <a class="reference external" href="https://www.gnu.org/software/make/manual/html_node/Functions.html">make function documentation</a> for an overview</p></li>
<li><p><a class="reference external" href="https://www.gnu.org/software/make/manual/html_node/Pattern-Rules.html">pattern rules</a> are great where possible, but sometimes are incompatible or just don’t fit
into my brain. feel free to swap them in where I’ve missed opportunities to do so</p></li>
<li><p>most pipelines begin with some form of enumeration of the possible outputs, and restriction
of those outputs to ones with valid corresponding inputs. a lot of this relies on <code class="docutils literal notranslate"><span class="pre">find</span></code>
to detect things</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">make</span></code> has no direct compatibility with <code class="docutils literal notranslate"><span class="pre">yaml</span></code>, which is a real pain. that’s been on
the to-do list for ages. thus the global configuration parameters in <code class="docutils literal notranslate"><span class="pre">Makefile.config</span></code> remain
in raw <code class="docutils literal notranslate"><span class="pre">make</span></code> instead of in a nice <code class="docutils literal notranslate"><span class="pre">yaml</span></code> file. please do patch in support as you feel like it</p></li>
<li><p>the most complex DAG construction happens with the analysis modules in <code class="docutils literal notranslate"><span class="pre">shared-makefiles/</span></code>; for that,
I built <a class="reference external" href="https://github.com/NCI-CGR/initialize_output_directories">a custom preprocessor</a> that actually deals with <code class="docutils literal notranslate"><span class="pre">yaml</span></code> natively and emits targets in a format
<code class="docutils literal notranslate"><span class="pre">make</span></code> can handle. see <code class="docutils literal notranslate"><span class="pre">shared-makefiles/Makefile.boltlmm</span></code> as an example.
my apologies that it’s not more generalized</p></li>
</ul>
</div>
<div class="section" id="job-tracking-logging">
<h3>Job Tracking/Logging<a class="headerlink" href="#job-tracking-logging" title="Permalink to this headline">¶</a></h3>
<p>All rules in the <code class="docutils literal notranslate"><span class="pre">make</span></code> pipelines are wrapped in utility functions that emit tracking files
(by default appended with <code class="docutils literal notranslate"><span class="pre">.success</span></code>, configurable with <code class="docutils literal notranslate"><span class="pre">$(TRACKING_SUCCESS_SUFFIX)</span></code>). Log data
are emitted to output files prefixed uniquely based on the rule. So everything is there, but it
can be a bit of a mess to find given how many files are emitted.</p>
</div>
<div class="section" id="integration-with-clusters">
<h3>Integration with Clusters<a class="headerlink" href="#integration-with-clusters" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">make</span></code> has no integrated support for cluster job submission. These pipelines have their calls wrapped
in two simple utility functions which either dispatch jobs via a <code class="docutils literal notranslate"><span class="pre">qsub</span></code> interface (for cgems), or run
the job in the main process but log the results with tracking files. This is a total mess. Other languages
obviously offer actual support, so further development should make use of that. For extension of the current
pipelines, support for <code class="docutils literal notranslate"><span class="pre">slurm</span></code> in particular needs to get patched into an appropriate interface/monitor program.</p>
</div>
<div class="section" id="conda-environments">
<h3>Conda Environments<a class="headerlink" href="#conda-environments" title="Permalink to this headline">¶</a></h3>
<p>Again, no intrinsic <code class="docutils literal notranslate"><span class="pre">make</span></code> support, and <code class="docutils literal notranslate"><span class="pre">make</span></code> runs things in subshells that make launching separate
environments rather complicated. It should be possible to wrap individual rules in activated <code class="docutils literal notranslate"><span class="pre">conda</span></code>
environments. However, given the comparative simplicity of the workspace as a whole, the top-level <code class="docutils literal notranslate"><span class="pre">conda</span></code>
environment should be sufficient for all modules, at least as currently written. Just make sure the top-level
environment is active before launching any step of the analysis.</p>
</div>
<div class="section" id="thread-safety-directory-locking">
<h3>Thread Safety/Directory Locking<a class="headerlink" href="#thread-safety-directory-locking" title="Permalink to this headline">¶</a></h3>
<p>Once again, <code class="docutils literal notranslate"><span class="pre">make</span></code> has no direct support for this, and I’ve never gotten around to actually implementing it.
This is probably the biggest silent weakness of the pipeline as written. Please keep this in mind when writing
and running pipelines. You can have bolt/saige/other analysis tools running at the same time; but don’t dispatch
multiple bolt jobs at the same time! It will collide with nasty consequences.</p>
</div>
</div>
<div class="section" id="adding-analysis-modules">
<h2>Adding Analysis Modules<a class="headerlink" href="#adding-analysis-modules" title="Permalink to this headline">¶</a></h2>
<p>In theory, it is very straightforward to add additional analysis
modules to the project by writing individual pipelines targeted
at specific tools (e.g. SNPTEST, PLINK), placing them in <code class="docutils literal notranslate"><span class="pre">shared-makefiles</span></code>,
and adding a global rule in <code class="docutils literal notranslate"><span class="pre">Makefile</span></code>. In practice, presumably no
future developers will want to write any such module in <code class="docutils literal notranslate"><span class="pre">make</span></code>, so
there need to be some thoughts to integrating modules from other languages.</p>
<p>I’m going to try to present the interface specification for these analysis
modules; with that, you should be able to write a module in any language,
and it should slot in pretty seamlessly.</p>
<div class="section" id="inputs">
<h3>Inputs<a class="headerlink" href="#inputs" title="Permalink to this headline">¶</a></h3>
<p>The following pertinent files will be available to the analysis pipeline
on run start. Note that all files are assumed to be prefixed with the installation
directory of the pipeline. Variables <code class="docutils literal notranslate"><span class="pre">$(VARIABLE_NAME)</span></code> are defined in <code class="docutils literal notranslate"><span class="pre">Makefile.config</span></code>.</p>
<ul class="simple">
<li><p>Chip data:</p>
<ul>
<li><p>each input platform and ancestry combination will have the following files,
so long as that combination has sufficient subjects for analysis:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">$(CLEANED_CHIP_OUTPUT_DIR)/{ancestry}/{platform}.step1.maf.geno.hwe.{bed,bim,fam}</span></code>: plink format genotypes with mild cleaning applied</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">$(CLEANED_CHIP_OUTPUT_DIR)/{ancestry}/{platform}.step2.pruning.{in,out}</span></code>: plink format output from –indep for variant {inclusion, exclusion}</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">$(CLEANED_CHIP_OUTPUT_DIR)/{ancestry}/{platform}.step3.pruning.{in,out}</span></code>: plink format output from –indep-pairwise for variant {inclusion, exclusion}</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">$(CLEANED_CHIP_OUTPUT_DIR)/{ancestry}/{platform}.step4.het.remove</span></code>: plink format subject removal file for heterozygosity outliers</p></li>
</ul>
</li>
</ul>
</li>
<li><p>Imputed data:</p>
<ul>
<li><p>each input platform and ancestry combination will have the following files,
so long as that combination has at least 100 subjects:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">$(BGEN_OUTPUT_DIR)/{platform}[/batch{#}]/{ancestry}/chr{#}-filtered.bgen</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">$(BGEN_OUTPUT_DIR)/{platform}[/batch{#}]/{ancestry}/chr{#}-filtered.bgen.bgi</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">$(BGEN_OUTPUT_DIR)/{platform}[/batch{#}]/{ancestry}/chr{#}-filtered-noNAs.sample</span></code></p></li>
</ul>
</li>
<li><p>Inputs are fixed at bgen version 1.2 dosages due to compatibility with bolt-lmm, saige, fastGWA</p></li>
<li><p>Modified sample file is due to format discrepancy in raw output from plink2</p></li>
</ul>
</li>
<li><p>Phenotype data:</p>
<ul>
<li><p>plain-text, tab-delimited phenotype file</p></li>
<li><p>header row</p></li>
<li><p>single column with unique subject ID (name is configurable as <code class="docutils literal notranslate"><span class="pre">$(PHENOTYPE_ID_COLNAME)</span></code>)</p></li>
<li><p>conventionally, missing values are stored as <code class="docutils literal notranslate"><span class="pre">NA</span></code></p></li>
</ul>
</li>
</ul>
</div>
<div class="section" id="outputs">
<h3>Outputs<a class="headerlink" href="#outputs" title="Permalink to this headline">¶</a></h3>
<ul>
<li><p>Results directory structure:</p>
<ul class="simple">
<li><p>all results should be emitted under $(RESULT_OUTPUT_DIR) as follows:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">$(RESULT_OUTPUT_DIR)/$(analysis_prefix)/{ancestry}/{software}</span></code></p></li>
<li><p>above, <code class="docutils literal notranslate"><span class="pre">$(analysis_prefix)</span></code> is the variable taken from <code class="docutils literal notranslate"><span class="pre">config/*.yaml</span></code></p></li>
</ul>
</li>
</ul>
</li>
<li><p>Output filenames:</p>
<ul class="simple">
<li><p>all intermediate and output files should be prefixed in the results directory as follows:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">$(phenotype).$(platform)[_batch{#}].{software}</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">$(phenotype)</span></code> is the variable taken from <code class="docutils literal notranslate"><span class="pre">config/*.yaml</span></code></p></li>
</ul>
</li>
</ul>
</li>
<li><p>Required output files and formats:</p>
<ul>
<li><p>the following files are those used downstream by existing pipeline components:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">$(phenotype).$(platform)[_batch{#}].{software}.tsv.gz</span></code></p>
<ul>
<li><p>results file per platform/batch</p></li>
<li><p>format is tab-delimited, columns as follows (with header as listed):</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">CHR</span></code>: chromosome of variant</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">POS</span></code>: physical position of variant, in GRCh38</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SNP</span></code>: variant ID (see note below)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Tested_Allele</span></code>: coded allele (corresponding to effect direction of BETA)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Other_Allele</span></code>: non-coded allele</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Freq_Tested_Allele_in_TOPMed</span></code>: allele frequency (see note below)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">BETA</span></code>: regression coefficient (binary traits: logOR) for variant</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SE</span></code>: standard error of test</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">P</span></code>: association p-value</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">N</span></code>: actual sample size tested for variant</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Ncases</span></code>: binary results only: actual number of cases tested for variant</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Ncontrols</span></code>: binary results only: actual number of controls tested for variant</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">SNP</span></code> defaults to “chr:pos:ref:alt” codes from TOPMed. This needs to be replaced
with rsIDs when requested with the <code class="docutils literal notranslate"><span class="pre">config/*.yaml</span></code> option <code class="docutils literal notranslate"><span class="pre">id_mode:</span> <span class="pre">rsid</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Freq_Tested_Allele_in_TOPMed</span></code> defaults to reference (approximate frequencies
from the imputation reference subjects) to avoid issues with identifiability of
subject samples. These should instead be replaced with actual subject allele
frequencies when requested with the <code class="docutils literal notranslate"><span class="pre">config/*.yaml</span></code> option <code class="docutils literal notranslate"><span class="pre">frequency_mode:</span> <span class="pre">subject</span></code>.</p></li>
</ul>
</li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">$(phenotype).$(platform)[_batch{#}].{software}.rawids.tsv</span></code></p>
<ul>
<li><p>the format of this file is the same as the above, except SNP must contain unique IDs,
in this case the “chr:pos:ref:alt” IDs from the TOPMed reference data</p></li>
<li><p>this file is canonically actually an upstream intermediate that leads to the above output file</p></li>
<li><p>note the lack of compression. this can be patched to behave differently</p></li>
<li><p>as things are currently configured, this file is required by <code class="docutils literal notranslate"><span class="pre">shared-makefiles/Makefile.metal</span></code>,
the meta-analysis pipeline. this is because the rsID mapping requested by <code class="docutils literal notranslate"><span class="pre">id_mode:</span> <span class="pre">rsid</span></code> and
used for the “Atlas” website creates duplicate sites in a very few cases, which causes
issues for <code class="docutils literal notranslate"><span class="pre">metal</span></code> when trying to unambiguously link variants to one another across platforms</p></li>
<li><p>this is an extremely messy behavior, and one I’d love to see patched out somehow in the future</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="adding-other-pipelines">
<h2>Adding Other Pipelines<a class="headerlink" href="#adding-other-pipelines" title="Permalink to this headline">¶</a></h2>
<p>In addition to the above, other pipelines will likely be needed if this project is to continue.
For example, <code class="docutils literal notranslate"><span class="pre">bgen</span></code> v1.2 format has worked well for the PLCO “Atlas” project, but will likely
need to be replaced or augmented in the future.</p>
<p>Most of the project’s pipelines live in a dedicated subdirectory of the appropriate name. They are
called from a dedicated rule in the top-level <code class="docutils literal notranslate"><span class="pre">Makefile</span></code>, and dispatch themselves based on variables
they import from <code class="docutils literal notranslate"><span class="pre">Makefile.config</span></code>. This process can be repeated for other necessary backend pipelines.</p>
<p>Note that, in particular for later pipelines operating on ancestry-split data, there needs to be
the capacity to dynamically restrict the DAG to combinations of platform and ancestry that exist
in the actual data, not just the full enumeration of platform and ancestry combinations. The <code class="docutils literal notranslate"><span class="pre">make</span></code>
pipelines do this by assuming upstream pipelines run to completion and detecting whatever output files
happen to be present from those pipelines, and working from there. Other languages have more elegant
support for this kind of DAG restriction. Just make sure you do it: there is never any guarantee
that any particular input combination will be present, and in fact for many ancestries given US sampling
criteria, it’s almost guaranteed they will be absent.</p>
<div class="section" id="extension-to-other-languages">
<h3>Extension to Other Languages<a class="headerlink" href="#extension-to-other-languages" title="Permalink to this headline">¶</a></h3>
<p>No one will want to write any further pipelines in <code class="docutils literal notranslate"><span class="pre">make</span></code>. However, it should be reasonably
straightforward to create modules in other languages. Make sure the modules conform to the above
interface specification, or possibly modify it while maintaining back compatibility.</p>
<p>The only major issue comes up around job dispatch. You can write a <code class="docutils literal notranslate"><span class="pre">snakemake</span></code> call into
the top-level <code class="docutils literal notranslate"><span class="pre">Makefile</span></code> dispatcher; however, that will not straightforwardly handle process
monitoring in the way recursive <code class="docutils literal notranslate"><span class="pre">make</span></code> usually does, and it loses out on a bunch of <code class="docutils literal notranslate"><span class="pre">snakemake</span></code>’s
convenient features.</p>
<p>The best solution then should be to create a language-specific dispatcher that handles module calls
within the language of the module. So, write a top-level <code class="docutils literal notranslate"><span class="pre">Snakefile</span></code> that covers <code class="docutils literal notranslate"><span class="pre">snakemake</span></code> analysis
modules. As ever, care must be taken to be sure upstream pipelines have run to completion before analysis.
However, the way the <code class="docutils literal notranslate"><span class="pre">make</span></code> pipelines are structured, that’s the case regardless, so the added
burden should be minimal.</p>
</div>
</div>
<div class="section" id="debugging">
<h2>Debugging<a class="headerlink" href="#debugging" title="Permalink to this headline">¶</a></h2>
<p>There are notes about this in the sections covering individual pipelines. However, I’ll list here
the biggest issues I’ve run into on a regular basis.</p>
<div class="section" id="wrong-environment-loaded">
<h3>Wrong Environment Loaded<a class="headerlink" href="#wrong-environment-loaded" title="Permalink to this headline">¶</a></h3>
<p>Obviously, the simplest way to check this is just glance at the active <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment
before dispatching jobs. But it’s easy to forget.</p>
<p>The most obvious issues that come up are as follows (and bear in mind most of the dev process
has been under different environments than these, so I’m still learning what the obvious issues
are):</p>
<ul class="simple">
<li><p>If you’re running <code class="docutils literal notranslate"><span class="pre">ldsc</span></code> or <code class="docutils literal notranslate"><span class="pre">ldscores</span></code> and incorrectly have <code class="docutils literal notranslate"><span class="pre">plco-analysis</span></code> active:
will report <code class="docutils literal notranslate"><span class="pre">ldsc.py</span></code> or <code class="docutils literal notranslate"><span class="pre">munge_sumstats.py</span></code> not available</p></li>
<li><p>If you’re running something other than <code class="docutils literal notranslate"><span class="pre">ldsc</span></code> or <code class="docutils literal notranslate"><span class="pre">ldscores</span></code> and have <code class="docutils literal notranslate"><span class="pre">plco-analysis-ldsc</span></code> active:
depends on which pipeline you’re running. The most likely issue will be the inability to find
some piece of software or reference data that’s cooked into the python3 conda environment. A short list of the
likeliest candidates:</p>
<ul>
<li><p>liftOver</p></li>
<li><p>plink2</p></li>
<li><p>bgenix</p></li>
<li><p>GRCh38 genetic map</p></li>
<li><p>any of the internal C++ programs ending in <code class="docutils literal notranslate"><span class="pre">.out</span></code> except for <code class="docutils literal notranslate"><span class="pre">qsub_job_monitor.out</span></code></p></li>
<li><p>graf, or its reference 1000 Genomes file <code class="docutils literal notranslate"><span class="pre">G1000FpGeno.bim</span></code></p></li>
<li><p>bolt or metal</p></li>
</ul>
</li>
</ul>
<p>Note that these behaviors are based on the basic installation landscape of cgems/ccad, so ymmv.</p>
</div>
<div class="section" id="problems-with-the-cluster">
<h3>Problems with the Cluster<a class="headerlink" href="#problems-with-the-cluster" title="Permalink to this headline">¶</a></h3>
<p>Speaking specifically of cgems/ccad and biowulf: if you run these pipelines enough times,
you <em>will</em> encounter issues with the cluster. The most common issues are as follows:</p>
<ul class="simple">
<li><p>Cluster non-responsiveness: failure to respond or dispatch</p></li>
<li><p>Desync between cluster memory writes and visible/accessible files</p></li>
<li><p>Stuck/dead nodes: job reports running but is in fact frozen and zombied</p></li>
</ul>
<p>Non-responsiveness isn’t always catastrophic. Small scale events don’t necessarily break
the pipeline: the qsub monitoring software has been designed to wait a number of intervals
between probing results, so if the event resolves itself shortly, the pipeline will continue
to function; and if it keeps going, the pipeline will not try to submit endlessly but instead
quit.</p>
<p>Desync is annoying but again, the qsub monitoring software has a series of retries to attempt
to allow for some amount of desync. The waiting times for this behavior are configurable, so
if you have issues, you can make the monitor (controlled in a macro in <code class="docutils literal notranslate"><span class="pre">Makefile.config</span></code>)
wait longer or retry more times to adapt. As it’s configured, I’ve not had any issues with cgems
in months.</p>
<p>Zombie jobs are obnoxious because it’s difficult to be certain when it’s happening. I am aware
that some pipelines at CGR deal with this by permitting a maximum amount of time between output
file updates before reporting an issue. This has not been such an issue that I’ve needed such a
failsafe, beyond merely checking <code class="docutils literal notranslate"><span class="pre">qstat</span></code> or <code class="docutils literal notranslate"><span class="pre">sjobs</span></code> periodically to see if all remaining jobs
are assigned to a suspiciously small number of nodes.</p>
</div>
<div class="section" id="genetic-heritability-near-zero">
<h3>Genetic Heritability Near Zero<a class="headerlink" href="#genetic-heritability-near-zero" title="Permalink to this headline">¶</a></h3>
<p>This is the biggest issue for all analysis tools. Each of the implementations in <code class="docutils literal notranslate"><span class="pre">shared-makefiles</span></code>
(bolt, fastGWA, saige) have some sort of issue if the phenotype model in question shows a near-zero
genetic heritability variance component, or an inflation near 1 equivalently.</p>
<p>This will manifest as failures in saige round 1; or in boltlmm during each imputed chromosome run
with a log message about trying standard linear models. There is not a good deal of sense around
these messages, in that sometimes low genetic heritability estimates seem to lead to run-ending
errors, and sometimes they don’t. It should be noted that this issue is very strongly correlated
with low sample size: small chip/ancestry combination, smaller heritability estimate. This also leads
to very (evidently) <em>unstable</em> heritability estimates. I will emphasize here that these programs were
<em>absolutely not</em> intended for use on sample sizes this low, so none of this behavior is unexpected,
nor does any of it constitute a bug, but rather a failure of study and analysis design.</p>
<p>How to find these issues:</p>
<ul class="simple">
<li><p>Run bolt or saige: <code class="docutils literal notranslate"><span class="pre">make</span> <span class="pre">boltlmm</span></code> or <code class="docutils literal notranslate"><span class="pre">make</span> <span class="pre">saige</span></code></p></li>
<li><p>Wait</p></li>
<li><p>Find errors in submission log indicating primary analysis rule failures</p></li>
<li><p>Check relevant information in <code class="docutils literal notranslate"><span class="pre">$(RESULTS_OUTPUT_DIR)</span></code>:</p>
<ul>
<li><p>for saige: <code class="docutils literal notranslate"><span class="pre">$(RESULTS_OUTPUT_DIR)/$(analysis_prefix)/$(ancestry)/SAIGE/$(phenotype).$(platform).saige.round1.varianceRatio.txt</span></code></p></li>
<li><p>for boltlmm: <code class="docutils literal notranslate"><span class="pre">$(RESULTS_OUTPUT_DIR)/$(analysis_prefix)/$(ancestry)/BOLTLMM/$(phenotype).$(platform).chr1.boltlmm.log</span></code></p></li>
</ul>
</li>
</ul>
<p>Solutions to these issues are limited. The most direct solution is <strong>remove the offending platform/ancestry combination from the configuration file</strong>.
No one likes this solution. But the alternatives are not very generalizable. One solution would be to have the investigator (if such
a person exists) remodel the trait in some fashion, possibly in a way that better captures a polygenic trait. The other possibility,
and this will be a bigger issue as people attempt to use these pipelines on other projects, is desyncing between phenotypes
and genotypes, possibly due to large-scale ID swaps. This could cause low apparent heritability that was in fact indicative
of dataset corruption.</p>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">plco-analysis</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Preprocessing.html">Preprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="Trait%20Configuration.html">Trait Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="Primary%20Analysis.html">Primary Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="Meta-Analysis.html">Meta-Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="Postprocessing.html">Postprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="Globus%20Distribution.html">Globus Distribution</a></li>
<li class="toctree-l1"><a class="reference internal" href="User%20Vignette.html">User Vignette</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Advanced Dev Notes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview-of-make">Overview of Make</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#rule-structures">Rule Structures</a></li>
<li class="toctree-l3"><a class="reference internal" href="#job-tracking-logging">Job Tracking/Logging</a></li>
<li class="toctree-l3"><a class="reference internal" href="#integration-with-clusters">Integration with Clusters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#conda-environments">Conda Environments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#thread-safety-directory-locking">Thread Safety/Directory Locking</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#adding-analysis-modules">Adding Analysis Modules</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#inputs">Inputs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#outputs">Outputs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#adding-other-pipelines">Adding Other Pipelines</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#extension-to-other-languages">Extension to Other Languages</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#debugging">Debugging</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#wrong-environment-loaded">Wrong Environment Loaded</a></li>
<li class="toctree-l3"><a class="reference internal" href="#problems-with-the-cluster">Problems with the Cluster</a></li>
<li class="toctree-l3"><a class="reference internal" href="#genetic-heritability-near-zero">Genetic Heritability Near Zero</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Methods%20Summary.html">Methods Summary</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="User%20Vignette.html" title="previous chapter">User Vignette</a></li>
      <li>Next: <a href="Methods%20Summary.html" title="next chapter">Methods Summary</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Lightning Auriga.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 3.4.3</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/Advanced Dev Notes.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>